File: web/src/lib/crypto.ts
// Client-side end-to-end encryption for chat messages. The server (see
// lib/api.ts's device key endpoints) only ever stores and relays public
// keys and ciphertext — it has no key material that could decrypt anything
// here, by construction.
//
// Scheme, per message:
//   1. Generate a random AES-256-GCM "content key" + IV, encrypt the text.
//   2. For every recipient device (every current member's every registered
//      device, plus the sender's own other devices), derive a per-device
//      wrapping key via ECDH(senderPriv, deviceRecipientPub) and use it to
//      wrap the content key with its own IV.
//   3. Ship ciphertext + iv + the array of per-device wrapped keys.
// Decrypting is the mirror: find your own device's entry, re-derive the
// same ECDH shared secret (ECDH is symmetric — recipientPriv+senderPub
// gives the same result), unwrap the content key, decrypt.
//
// This has no ratcheting/forward-secrecy beyond "one fresh AES key per
// message" — a real Signal-style double-ratchet is out of scope here.
import { deviceID } from "./ws";
import { api } from "./api";

const DB_NAME = "vc-e2e-keys";
const STORE = "keypair";
// Keypairs are per account (record "device:<userID>"), not per browser: on a
// shared machine the next person to sign in must not inherit — and publish
// under their own account — the previous user's private key. "device" is
// the old single, browser-wide record, adopted once by its real owner.
const LEGACY_RECORD_KEY = "device";
const recordKey = (userID: number) => `device:${userID}`;

interface StoredKeyPair {
  publicKey: CryptoKey;
  privateKey: CryptoKey;
}

interface WrappedKeyEntry {
  user_id: number;
  device_id: string;
  wrapped_key: string; // base64
  wrap_iv: string; // base64
  sender_pub_jwk: JsonWebKey;
}

function openKeyDB(): Promise<IDBDatabase> {
  return new Promise((resolve, reject) => {
    const req = indexedDB.open(DB_NAME, 1);
    req.onupgradeneeded = () => req.result.createObjectStore(STORE);
    req.onsuccess = () => resolve(req.result);
    req.onerror = () => reject(req.error);
  });
}

async function idbGet<T>(store: string, key: string): Promise<T | undefined> {
  const db = await openKeyDB();
  return new Promise((resolve, reject) => {
    const tx = db.transaction(store, "readonly").objectStore(store).get(key);
    tx.onsuccess = () => resolve(tx.result);
    tx.onerror = () => reject(tx.error);
  });
}

async function idbDelete(store: string, key: string): Promise<void> {
  const db = await openKeyDB();
  return new Promise((resolve, reject) => {
    const tx = db.transaction(store, "readwrite").objectStore(store).delete(key);
    tx.onsuccess = () => resolve();
    tx.onerror = () => reject(tx.error);
  });
}

async function idbSet(store: string, key: string, value: unknown): Promise<void> {
  const db = await openKeyDB();
  return new Promise((resolve, reject) => {
    const tx = db.transaction(store, "readwrite").objectStore(store).put(value, key);
    tx.onsuccess = () => resolve();
    tx.onerror = () => reject(tx.error);
  });
}

let cachedKeyPair: StoredKeyPair | null = null;
let registerPromise: Promise<void> | null = null;
let currentUserID: number | null = null;

// adoptLegacyKeyPair returns the old browser-wide keypair only if the server
// has it registered as this user's key for this device — i.e. it really was
// theirs — then moves it to the per-account record.
async function adoptLegacyKeyPair(userID: number): Promise<StoredKeyPair | undefined> {
  const legacy = await idbGet<StoredKeyPair>(STORE, LEGACY_RECORD_KEY);
  if (!legacy) return undefined;
  try {
    const pub = await crypto.subtle.exportKey("jwk", legacy.publicKey);
    const mine = (await api.deviceKeys([userID])).find((k) => k.device_id === deviceID());
    const registered = mine ? (JSON.parse(mine.public_key_jwk) as JsonWebKey) : null;
    if (!registered || registered.x !== pub.x || registered.y !== pub.y) return undefined;
  } catch {
    return undefined;
  }
  await idbSet(STORE, recordKey(userID), legacy);
  await idbDelete(STORE, LEGACY_RECORD_KEY);
  return legacy;
}

export function isCryptoSubtleAvailable(): boolean {
  return typeof crypto !== "undefined" && !!crypto.subtle;
}

// getDeviceKeyPair returns this browser's ECDH keypair, generating one
// (non-extractable private key — it can never be read back out, only used)
// the first time and persisting it in IndexedDB from then on.
async function getDeviceKeyPair(): Promise<StoredKeyPair> {
  if (!isCryptoSubtleAvailable()) {
    throw new Error("WebCrypto is unavailable in insecure HTTP contexts. Connect over HTTPS to enable encryption.");
  }
  if (cachedKeyPair) return cachedKeyPair;
  const userID = currentUserID;
  if (userID == null) throw new Error("E2E keys are not available before sign-in");
  const stored = (await idbGet<StoredKeyPair>(STORE, recordKey(userID))) ?? (await adoptLegacyKeyPair(userID));
  if (stored) {
    cachedKeyPair = stored;
    return stored;
  }
  const pair = (await crypto.subtle.generateKey({ name: "ECDH", namedCurve: "P-256" }, false, [
    "deriveKey",
  ])) as CryptoKeyPair;
  // The public key must be extractable to export as JWK for the server;
  // generateKey's `extractable` flag applies to both halves of the pair, so
  // re-derive an extractable public key by exporting/reimporting it — the
  // private key from generateKey (extractable: false) never had that option.
  const rawPub = await crypto.subtle.exportKey("jwk", pair.publicKey);
  const extractablePub = await crypto.subtle.importKey(
    "jwk", rawPub, { name: "ECDH", namedCurve: "P-256" }, true, [],
  );
  const record: StoredKeyPair = { publicKey: extractablePub, privateKey: pair.privateKey };
  await idbSet(STORE, recordKey(userID), record);
  cachedKeyPair = record;
  return record;
}

// ensureRegistered publishes this device's public key to the server, once
// per app load. Safe to call repeatedly (idempotent server-side upsert).
export function ensureDeviceRegistered(userID: number): Promise<void> {
  if (!isCryptoSubtleAvailable()) {
    return Promise.resolve();
  }
  if (currentUserID !== userID) {
    currentUserID = userID;
    cachedKeyPair = null;
    registerPromise = null;
  }
  if (!registerPromise) {
    registerPromise = (async () => {
      const { publicKey } = await getDeviceKeyPair();
      const jwk = await crypto.subtle.exportKey("jwk", publicKey);
      await api.registerDeviceKey(deviceID(), JSON.stringify(jwk));
    })();
  }
  return registerPromise;
}

function b64encode(buf: ArrayBuffer): string {
  const bytes = new Uint8Array(buf);
  let binary = "";
  const chunkSize = 8192;
  for (let i = 0; i < bytes.length; i += chunkSize) {
    binary += String.fromCharCode(...bytes.subarray(i, i + chunkSize));
  }
  return btoa(binary);
}
function b64decode(b64: string): Uint8Array {
  return Uint8Array.from(atob(b64), (c) => c.charCodeAt(0));
}

async function deriveWrapKey(privateKey: CryptoKey, peerPublicJwk: JsonWebKey): Promise<CryptoKey> {
  const peerPub = await crypto.subtle.importKey("jwk", peerPublicJwk, { name: "ECDH", namedCurve: "P-256" }, false, []);
  return crypto.subtle.deriveKey(
    { name: "ECDH", public: peerPub }, privateKey, { name: "AES-GCM", length: 256 }, false, ["encrypt", "decrypt"],
  );
}

export interface EncryptedPayload {
  ciphertext: string;
  iv: string;
  encKeys: string; // JSON-stringified WrappedKeyEntry[]
}

// canEncryptTo checks whether every target user has at least one registered
// device key — the gate for showing/using the "encrypt this conversation"
// toggle. Returns the userIDs that are NOT ready, for a useful error.
export async function usersMissingKeys(userIDs: number[]): Promise<number[]> {
  const keys = await api.deviceKeys(userIDs);
  const ready = new Set(keys.map((k) => k.user_id));
  return userIDs.filter((id) => !ready.has(id));
}

// encryptForRecipients encrypts plaintext for every device of every user in
// recipientUserIDs (which should include the sender's own id, so their
// other devices/history sync too — callers pass the full participant set).
export async function encryptForRecipients(plaintext: string, recipientUserIDs: number[]): Promise<EncryptedPayload> {
  const { publicKey, privateKey } = await getDeviceKeyPair();
  const ownPubJwk = await crypto.subtle.exportKey("jwk", publicKey);

  const allKeys = await api.deviceKeys(recipientUserIDs);
  if (allKeys.length === 0) {
    throw new Error("No recipient has encryption set up on any device yet");
  }

  const contentKey = await crypto.subtle.generateKey({ name: "AES-GCM", length: 256 }, true, ["encrypt", "decrypt"]);
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const ciphertextBuf = await crypto.subtle.encrypt({ name: "AES-GCM", iv }, contentKey, new TextEncoder().encode(plaintext));
  const rawContentKey = new Uint8Array(await crypto.subtle.exportKey("raw", contentKey));

  // Every device gets a wrapped key, including the sender's own current
  // one — reloading the page (or another tab) later needs to re-derive the
  // plaintext from stored history same as any other device would.
  const encKeys: WrappedKeyEntry[] = [];
  for (const dk of allKeys) {
    try {
      const wrapKey = await deriveWrapKey(privateKey, JSON.parse(dk.public_key_jwk));
      const wrapIv = crypto.getRandomValues(new Uint8Array(12));
      const wrapped = await crypto.subtle.encrypt({ name: "AES-GCM", iv: wrapIv }, wrapKey, rawContentKey);
      encKeys.push({
        user_id: dk.user_id, device_id: dk.device_id,
        wrapped_key: b64encode(wrapped), wrap_iv: b64encode(wrapIv.buffer as ArrayBuffer),
        sender_pub_jwk: ownPubJwk as JsonWebKey,
      });
    } catch {
      // A malformed/incompatible key on one device shouldn't sink the whole send.
    }
  }

  return {
    ciphertext: b64encode(ciphertextBuf),
    iv: b64encode(iv.buffer as ArrayBuffer),
    encKeys: JSON.stringify(encKeys),
  };
}

const MAX_DECRYPT_CACHE = 500;
const decryptCache = new Map<number, string>();

// Drop decrypted plaintext held in memory when the active account changes —
// this app is designed to run on shared/kiosk-style devices, so decrypted
// content from a previous session must not linger in memory indefinitely.
export function clearDecryptCache(): void {
  decryptCache.clear();
  cachedKeyPair = null;
  registerPromise = null;
  currentUserID = null;
}

// decryptMessageContent returns the plaintext for an encrypted message on
// this device, or null if this device has no wrapped key for it (sent
// before this device registered, or the sender excluded it). Results are
// cached by message id since decryption is a few async crypto calls.
export async function decryptMessageContent(msg: { id: number; is_encrypted: boolean; content: string; enc_iv?: string | null; enc_keys?: string | null }): Promise<string | null> {
  if (!msg.is_encrypted) return msg.content;
  if (!isCryptoSubtleAvailable()) return "[Encrypted message — HTTPS required to decrypt]";
  if (decryptCache.has(msg.id)) {
    const val = decryptCache.get(msg.id)!;
    decryptCache.delete(msg.id);
    decryptCache.set(msg.id, val);
    return val;
  }
  const result = await decryptNow(msg);
  if (result !== null) {
    if (decryptCache.size >= MAX_DECRYPT_CACHE) {
      const oldestKey = decryptCache.keys().next().value;
      if (oldestKey !== undefined) {
        decryptCache.delete(oldestKey);
      }
    }
    decryptCache.set(msg.id, result);
  }
  return result;
}

async function decryptNow(msg: { content: string; enc_iv?: string | null; enc_keys?: string | null }): Promise<string | null> {
  if (!msg.enc_keys || !msg.enc_iv) return null;
  let entries: WrappedKeyEntry[];
  try {
    entries = JSON.parse(msg.enc_keys);
  } catch {
    return null;
  }
  const myDeviceId = deviceID();
  const mine = entries.find((e) => e.device_id === myDeviceId);
  if (!mine) return null;
  try {
    const { privateKey } = await getDeviceKeyPair();
    const wrapKey = await deriveWrapKey(privateKey, mine.sender_pub_jwk);
    const rawContentKey = await crypto.subtle.decrypt({ name: "AES-GCM", iv: b64decode(mine.wrap_iv) }, wrapKey, b64decode(mine.wrapped_key));
    const contentKey = await crypto.subtle.importKey("raw", rawContentKey, { name: "AES-GCM" }, false, ["decrypt"]);
    const plainBuf = await crypto.subtle.decrypt({ name: "AES-GCM", iv: b64decode(msg.enc_iv) }, contentKey, b64decode(msg.content));
    return new TextDecoder().decode(plainBuf);
  } catch {
    return null;
  }
}

File: android/app/src/main/java/com/videocall/mobile/chat/Crypto.kt
package com.videocall.mobile.chat

import android.content.Context
import android.security.keystore.KeyProperties
import android.util.Base64
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.decodeFromJsonElement
import java.math.BigInteger
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.PrivateKey
import java.security.PublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import com.videocall.mobile.net.DeviceKey
import com.videocall.mobile.net.Prefs
import com.videocall.mobile.session.SessionManager

/**
 * Client-side E2E chat encryption — a straight port of web/src/lib/crypto.ts
 * to the JVM: ECDH P-256 per-device keypair, AES-256-GCM content key per
 * message, the content key wrapped per-recipient-device via
 * ECDH(myPriv, theirPub). The server only ever sees public keys + ciphertext
 * (see api.registerDeviceKey/deviceKeys). No ratcheting/forward-secrecy
 * beyond "one fresh AES key per message", same as the web client.
 *
 * The private key is generated on-device and stored in EncryptedSharedPreferences
 * (an Android-Keystore-wrapped file) — it never leaves the device and is
 * never sent anywhere; only the public key is uploaded.
 */
object Crypto {
    private const val CURVE = "secp256r1" // NIST P-256, same curve WebCrypto calls "P-256"
    private const val PREFS = "vc_e2e_keys"
    // Keys are stored per account ("priv:<userId>"): on a shared phone the
    // next person to sign in must not inherit — and publish under their own
    // account — the previous user's private key. The unsuffixed names are the
    // old device-wide record, adopted once by its real owner (see below).
    private const val LEGACY_PRIV = "priv"
    private const val LEGACY_PUB = "pub_jwk"
    private fun privKey(userId: Long) = "priv:$userId"
    private fun pubKey(userId: Long) = "pub_jwk:$userId"

    @kotlinx.serialization.Serializable
    private data class WrappedKeyEntry(
        val user_id: Long,
        val device_id: String,
        val wrapped_key: String,
        val wrap_iv: String,
        val sender_pub_jwk: JsonObject,
    )

    data class KeyPairBundle(val private: PrivateKey, val public: PublicKey, val publicJwk: JsonObject)

    private var cached: KeyPairBundle? = null
    private var cachedFor: Long? = null

    private fun prefs(context: Context) = EncryptedSharedPreferences.create(
        context,
        PREFS,
        MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
    )

    private fun currentUserId(): Long =
        SessionManager.me.value?.id ?: throw IllegalStateException("E2E keys are not available before sign-in")

    private suspend fun getOrCreateKeyPair(context: Context): KeyPairBundle {
        val userId = currentUserId()
        if (cachedFor == userId) cached?.let { return it }
        val p = prefs(context)
        val kf = KeyFactory.getInstance("EC")
        if (!p.contains(privKey(userId))) adoptLegacyKeyPair(context, userId)
        val storedPriv = p.getString(privKey(userId), null)
        val storedPub = p.getString(pubKey(userId), null)
        val bundle = if (storedPriv != null && storedPub != null) {
            val priv = kf.generatePrivate(PKCS8EncodedKeySpec(Base64.decode(storedPriv, Base64.NO_WRAP)))
            val jwk = json.parseToJsonElement(storedPub) as JsonObject
            KeyPairBundle(priv, publicKeyFromJwk(jwk), jwk)
        } else {
            val gen = KeyPairGenerator.getInstance("EC").apply { initialize(ECGenParameterSpec(CURVE)) }
            val pair = gen.generateKeyPair()
            val jwk = publicKeyToJwk(pair.public as java.security.interfaces.ECPublicKey)
            p.edit()
                .putString(privKey(userId), Base64.encodeToString(pair.private.encoded, Base64.NO_WRAP))
                .putString(pubKey(userId), jwk.toString())
                .apply()
            KeyPairBundle(pair.private, pair.public, jwk)
        }
        cached = bundle
        cachedFor = userId
        return bundle
    }

    /**
     * Moves the old device-wide keypair to [userId]'s record — but only if the
     * server has exactly that public key registered as this user's key for this
     * device, i.e. it really was theirs. Anyone else gets a fresh keypair.
     */
    private suspend fun adoptLegacyKeyPair(context: Context, userId: Long) {
        val p = prefs(context)
        val legacyPriv = p.getString(LEGACY_PRIV, null) ?: return
        val legacyPub = p.getString(LEGACY_PUB, null) ?: return
        val registered = runCatching { SessionManager.api.deviceKeys(listOf(userId)) }.getOrNull()
            ?.firstOrNull { it.device_id == Prefs.deviceId(context) } ?: return
        val mine = runCatching { json.parseToJsonElement(registered.public_key_jwk) as JsonObject }.getOrNull() ?: return
        val legacy = json.parseToJsonElement(legacyPub) as JsonObject
        if (mine["x"] != legacy["x"] || mine["y"] != legacy["y"]) return
        p.edit()
            .putString(privKey(userId), legacyPriv)
            .putString(pubKey(userId), legacyPub)
            .remove(LEGACY_PRIV)
            .remove(LEGACY_PUB)
            .apply()
    }

    /**
     * Publishes this device's public key for the signed-in account. Must run
     * after login (the endpoint needs a session) — calling it at app start
     * used to 401 silently on a fresh login, so nobody could encrypt to this
     * phone until the app was restarted.
     */
    suspend fun ensureDeviceRegistered(context: Context) {
        val bundle = runCatching { getOrCreateKeyPair(context) }.getOrNull() ?: return
        runCatching { SessionManager.api.registerDeviceKey(Prefs.deviceId(context), bundle.publicJwk.toString()) }
    }

    private fun fieldElement(n: BigInteger, size: Int = 32): ByteArray {
        val raw = n.toByteArray()
        val trimmed = if (raw.size > size) raw.copyOfRange(raw.size - size, raw.size) else raw
        if (trimmed.size == size) return trimmed
        val out = ByteArray(size)
        System.arraycopy(trimmed, 0, out, size - trimmed.size, trimmed.size)
        return out
    }

    private fun b64url(bytes: ByteArray): String =
        Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)

    private fun b64urlDecode(s: String): ByteArray = Base64.decode(s, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)

    private fun publicKeyToJwk(pub: java.security.interfaces.ECPublicKey): JsonObject = JsonObject(mapOf(
        "kty" to JsonPrimitive("EC"),
        "crv" to JsonPrimitive("P-256"),
        "x" to JsonPrimitive(b64url(fieldElement(pub.w.affineX))),
        "y" to JsonPrimitive(b64url(fieldElement(pub.w.affineY))),
        "ext" to JsonPrimitive(true),
    ))

    private val p256Params by lazy {
        val params = java.security.AlgorithmParameters.getInstance("EC")
        params.init(ECGenParameterSpec(CURVE))
        params.getParameterSpec(java.security.spec.ECParameterSpec::class.java)
    }

    private fun publicKeyFromJwk(jwk: JsonObject): PublicKey {
        val x = BigInteger(1, b64urlDecode((jwk["x"] as JsonPrimitive).content))
        val y = BigInteger(1, b64urlDecode((jwk["y"] as JsonPrimitive).content))
        val point = ECPoint(x, y)
        val spec = ECPublicKeySpec(point, p256Params)
        return KeyFactory.getInstance("EC").generatePublic(spec)
    }

    private fun deriveWrapKey(myPriv: PrivateKey, peerPub: PublicKey): SecretKeySpec {
        val ka = KeyAgreement.getInstance("ECDH")
        ka.init(myPriv)
        ka.doPhase(peerPub, true)
        val secret = ka.generateSecret() // 32 bytes for P-256, same bytes WebCrypto's ECDH deriveKey uses directly
        return SecretKeySpec(secret, "AES")
    }

    /** userIDs that do NOT have any registered device key yet — the gate for offering encryption. */
    suspend fun usersMissingKeys(userIds: List<Long>): List<Long> {
        val keys = runCatching { SessionManager.api.deviceKeys(userIds) }.getOrDefault(emptyList())
        val ready = keys.map { it.user_id }.toSet()
        return userIds.filterNot { it in ready }
    }

    data class EncryptedPayload(val ciphertext: String, val iv: String, val encKeys: String)

    suspend fun encryptForRecipients(context: Context, plaintext: String, recipientUserIds: List<Long>): EncryptedPayload {
        val bundle = getOrCreateKeyPair(context)
        val allKeys = SessionManager.api.deviceKeys(recipientUserIds)
        if (allKeys.isEmpty()) throw IllegalStateException("No recipient has encryption set up on any device yet")

        val contentKeyBytes = ByteArray(32).also { java.security.SecureRandom().nextBytes(it) }
        val contentKey = SecretKeySpec(contentKeyBytes, "AES")
        val iv = ByteArray(12).also { java.security.SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, contentKey, GCMParameterSpec(128, iv))
        val ciphertext = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))

        val entries = mutableListOf<WrappedKeyEntry>()
        for (dk: DeviceKey in allKeys) {
            try {
                val peerJwk = json.parseToJsonElement(dk.public_key_jwk) as JsonObject
                val peerPub = publicKeyFromJwk(peerJwk)
                val wrapKey = deriveWrapKey(bundle.private, peerPub)
                val wrapIv = ByteArray(12).also { java.security.SecureRandom().nextBytes(it) }
                val wrapCipher = Cipher.getInstance("AES/GCM/NoPadding")
                wrapCipher.init(Cipher.ENCRYPT_MODE, wrapKey, GCMParameterSpec(128, wrapIv))
                val wrapped = wrapCipher.doFinal(contentKeyBytes)
                entries += WrappedKeyEntry(
                    user_id = dk.user_id, device_id = dk.device_id,
                    wrapped_key = b64std(wrapped), wrap_iv = b64std(wrapIv),
                    sender_pub_jwk = bundle.publicJwk,
                )
            } catch (_: Exception) {
                // a malformed/incompatible key on one device shouldn't sink the whole send
            }
        }
        return EncryptedPayload(
            ciphertext = b64std(ciphertext),
            iv = b64std(iv),
            encKeys = json.encodeToString(ListSerializer(WrappedKeyEntry.serializer()), entries),
        )
    }

    private fun b64std(bytes: ByteArray): String = Base64.encodeToString(bytes, Base64.NO_WRAP)
    private fun b64stdDecode(s: String): ByteArray = Base64.decode(s, Base64.NO_WRAP)

    private val decryptCache = mutableMapOf<Long, String?>()

    /** Drop decrypted plaintext held in memory when the active account changes. */
    fun clearOnLogout() {
        decryptCache.clear()
        cached = null
        cachedFor = null
    }

    suspend fun decryptMessageContent(context: Context, id: Long, isEncrypted: Boolean, content: String, encIv: String?, encKeys: String?): String? {
        if (!isEncrypted) return content
        decryptCache[id]?.let { return it }
        if (!decryptCache.containsKey(id)) {
            val result = runCatching { decryptNow(context, content, encIv, encKeys) }.getOrNull()
            decryptCache[id] = result
            return result
        }
        return null
    }

    private suspend fun decryptNow(context: Context, content: String, encIv: String?, encKeysJson: String?): String? {
        if (encIv == null || encKeysJson == null) return null
        val entries = runCatching { json.decodeFromString<List<WrappedKeyEntry>>(encKeysJson) }.getOrNull() ?: return null
        val myDeviceId = Prefs.deviceId(context)
        // Sealed for this device, or else for the account's key backup (a
        // message from before this device existed, readable after restoring).
        val own = entries.firstOrNull { it.device_id == myDeviceId }
        val mine: WrappedKeyEntry
        val myPriv: PrivateKey
        if (own != null) {
            mine = own
            myPriv = getOrCreateKeyPair(context).private
        } else {
            mine = entries.firstOrNull { it.device_id == BACKUP_DEVICE && it.user_id == currentUserId() } ?: return null
            myPriv = backupPrivateKey(context) ?: return null
        }
        val senderPub = publicKeyFromJwk(mine.sender_pub_jwk)
        val wrapKey = deriveWrapKey(myPriv, senderPub)
        val unwrapCipher = Cipher.getInstance("AES/GCM/NoPadding")
        unwrapCipher.init(Cipher.DECRYPT_MODE, wrapKey, GCMParameterSpec(128, b64stdDecode(mine.wrap_iv)))
        val rawContentKey = unwrapCipher.doFinal(b64stdDecode(mine.wrapped_key))
        val contentKey = SecretKeySpec(rawContentKey, "AES")
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, contentKey, GCMParameterSpec(128, b64stdDecode(encIv)))
        val plain = cipher.doFinal(b64stdDecode(content))
        return String(plain, Charsets.UTF_8)
    }

    private val json = Json { ignoreUnknownKeys = true }

    // ---- account key backup ----
    //
    // Same scheme and format as the web app (web/src/lib/crypto.ts): an extra
    // "backup" key pair whose public key is registered like another device of
    // the user, so new messages are sealed for it too. Its private key is kept
    // on the server encrypted with a password (PBKDF2-SHA256 + AES-GCM).

    private const val BACKUP_DEVICE = "backup"
    private const val PBKDF2_ITERATIONS = 310_000
    private fun backupPrivPref(userId: Long) = "backup_priv:$userId"

    private fun backupPrivateKey(context: Context): PrivateKey? {
        val stored = prefs(context).getString(backupPrivPref(currentUserId()), null) ?: return null
        return runCatching { KeyFactory.getInstance("EC").generatePrivate(PKCS8EncodedKeySpec(b64stdDecode(stored))) }.getOrNull()
    }

    /** Whether this device can already read messages sealed for the backup. */
    fun hasBackupKey(context: Context): Boolean = prefs(context).contains(backupPrivPref(currentUserId()))

    private fun passwordKey(password: String, salt: ByteArray, iterations: Int): SecretKeySpec {
        val factory = javax.crypto.SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val spec = javax.crypto.spec.PBEKeySpec(password.toCharArray(), salt, iterations, 256)
        return SecretKeySpec(factory.generateSecret(spec).encoded, "AES")
    }

    private fun seal(pkcs8: ByteArray, password: String): JsonObject {
        val random = java.security.SecureRandom()
        val salt = ByteArray(16).also { random.nextBytes(it) }
        val iv = ByteArray(12).also { random.nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, passwordKey(password, salt, PBKDF2_ITERATIONS), GCMParameterSpec(128, iv))
        return JsonObject(mapOf(
            "v" to JsonPrimitive(1),
            "kdf" to JsonPrimitive("PBKDF2-SHA256"),
            "iterations" to JsonPrimitive(PBKDF2_ITERATIONS),
            "salt" to JsonPrimitive(b64std(salt)),
            "iv" to JsonPrimitive(b64std(iv)),
            "ciphertext" to JsonPrimitive(b64std(cipher.doFinal(pkcs8))),
        ))
    }

    private fun storeBackupKey(context: Context, pkcs8: ByteArray) {
        prefs(context).edit().putString(backupPrivPref(currentUserId()), b64std(pkcs8)).apply()
        decryptCache.clear()
    }

    /** Turns on backup, or replaces it with a new key and password. */
    suspend fun createKeyBackup(context: Context, password: String) = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.Default) {
        val pair = KeyPairGenerator.getInstance("EC").apply { initialize(ECGenParameterSpec(CURVE)) }.generateKeyPair()
        val jwk = publicKeyToJwk(pair.public as java.security.interfaces.ECPublicKey)
        SessionManager.api.saveKeyBackup(seal(pair.private.encoded, password), jwk.toString())
        storeBackupKey(context, pair.private.encoded)
    }

    /** Re-encrypts the existing backup key with a new password. */
    suspend fun changeKeyBackupPassword(context: Context, password: String) = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.Default) {
        val stored = prefs(context).getString(backupPrivPref(currentUserId()), null)
            ?: throw IllegalStateException("Restore the backup on this device first")
        val pub = SessionManager.api.deviceKeys(listOf(currentUserId())).firstOrNull { it.device_id == BACKUP_DEVICE }
            ?: throw IllegalStateException("There is no backup to update")
        SessionManager.api.saveKeyBackup(seal(b64stdDecode(stored), password), pub.public_key_jwk)
    }

    /** Unlocks the account's backup on this device with its password. */
    suspend fun restoreKeyBackup(context: Context, password: String) = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.Default) {
        val data = SessionManager.api.keyBackup().data ?: throw IllegalStateException("There is no backup for this account")
        fun field(name: String) = (data[name] as? JsonPrimitive)?.content ?: throw IllegalStateException("The backup is damaged")
        val iterations = field("iterations").toInt()
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, passwordKey(password, b64stdDecode(field("salt")), iterations), GCMParameterSpec(128, b64stdDecode(field("iv"))))
        val pkcs8 = try {
            cipher.doFinal(b64stdDecode(field("ciphertext")))
        } catch (e: javax.crypto.AEADBadTagException) {
            throw IllegalArgumentException("That password doesn't match the backup")
        }
        KeyFactory.getInstance("EC").generatePrivate(PKCS8EncodedKeySpec(pkcs8)) // must be a valid key
        storeBackupKey(context, pkcs8)
    }

    suspend fun deleteKeyBackup(context: Context) {
        SessionManager.api.deleteKeyBackup()
        prefs(context).edit().remove(backupPrivPref(currentUserId())).apply()
    }
}
